Place default ID attribute in Read Only group

nmuesch opened the pull request:
Fixes #45 

If the user hasn't specified the "ID" attribute in their schema, move it to the "Read Only" group so its not assumed to be a field that can be set explicitly in the terraform config.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-plugin-docs?pullRequest=46)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-plugin-docs?pullRequest=46)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-plugin-docs?pullRequest=46).</sub>

detro commented on schemamd/render.go:
```diff
@@ -161,7 +158,13 @@ func writeBlockChildren(w io.Writer, parents []string, block *tfjson.SchemaBlock
 		childAtt := block.Attributes[n]
 		for i, gf := range groupFilters {
-			if gf.filter(childBlock, childAtt) {
+			if n == "id" && childAtt.Description == "" {
```
Nit: I'd do a lowercase comparison with `id`, just to stay on the safe side

detro commented on schemamd/render.go:
```diff
@@ -161,7 +158,13 @@ func writeBlockChildren(w io.Writer, parents []string, block *tfjson.SchemaBlock
 		for i, gf := range groupFilters {
-			if gf.filter(childBlock, childAtt) {
+			if n == "id" && childAtt.Description == "" {
+				if gf.name == "Read Only" {
```
Instead of creating a new field `name` for `group filter`, just to support this comparison, wouldn't it be easier to just do a string contains `Read-Only`?

detro reviewed (changes requested):
This is a nice improvement to an annoying problem.

Will likely need to land in 0.8.0. Can you please add an entry in the CHANGELOG?

nmuesch commented:
Hey @detro Thanks for the review, I brought this up to date with the main branch and believe I've addressed your comments. 

I haven't been working on terraform in a while though, so while I tried to do a quick sanity check that this still works as expected, it'd be great if someone were able to give this a test on their side.

detro commented on schemamd/render.go:
```diff
@@ -225,6 +225,13 @@ nameLoop:
 		} else if childAtt, ok := block.Attributes[n]; ok {
 			for i, gf := range groupFilters {
+				if strings.ToLower(n) == "id" && childAtt.Description == "" {
+					if gf.topLevelTitle == "Read-Only" {
```
Did you verify this? Because I think this won't match anything: I believe the string is `### Read-Only`.

It would be even better if the elements of `groupFilters` above existed as variables themselves, and so you could directly match that and avoid bugs out of string mismatches (now and in the future).

OJFord replied:
That's correct, I have it working with that change (see below).

detro commented on schemamd/render.go:
```diff
@@ -225,6 +225,13 @@ nameLoop:
 			for i, gf := range groupFilters {
+				if strings.ToLower(n) == "id" && childAtt.Description == "" {
+					if gf.topLevelTitle == "Read-Only" {
+						childAtt.Description = "The ID of this resource."
```
I noticed you are not removing anymore the `if` [here](https://github.com/hashicorp/terraform-plugin-docs/pull/46/files#diff-e41a127c178c2eabe9f50f358b74b7c81df61ca60719d28f66a698009fd43b4cR73) that sets the description for `id`. What changed?

OJFord replied:
Yes it's also necessary to remove that, and I'm not sure it's worth checking for empty description here anyway?

detro replied:
If we don't do that, all descriptions currently set on `ID`s will be overridden by this change, making it impossible for users to tailor-make the description they want.

OJFord replied:
But then only the description setting should be inside this condition, not the appending to Read-Only group too?

detro commented on CHANGELOG.md:
```diff
@@ -3,7 +3,11 @@
+
+BUG FIXES:
+
+* schemamd: Place the default "ID" attribute as a "Read Only" attribute, unless there's a defined description ([#46](https://github.com/hashicorp/terraform-plugin-docs/pull/46)).
```
```suggestion
* schemamd: By default `ID` attribute should be "Read Only", unless there's a description defined, in which case it's handled like any other attribute in the schema ([#46](https://github.com/hashicorp/terraform-plugin-docs/pull/46)).
```

detro reviewed (changes requested):
I think I might have identified an issue, and also provided some suggestions.

Also, please ensure you have tried these changes, as the testing facilities in this repo are a bit limited, so manual verification is key.

OJFord commented:
> Also, please ensure you have tried these changes, as the testing facilities in this repo are a bit limited, so manual verification is key.

> I haven't been working on terraform in a while though, so while I tried to do a quick sanity check that this still works as expected, it'd be great if someone were able to give this a test on their side.

Hi, many thanks for working on this @nmuesch! I found your original report after a user report in OJFord/terraform-provider-wireguard#8.

I've just tested your branch against that repo:
```
$ (cd terraform-plugin-docs && go build cmd/tfplugindocs)
$ (cd terraform-provider-wireguard && rm -r docs && ../terraform-plugin-docs/tfplugindocs generate && git status)
```

and unfortunately the docs are generated exactly as they were by v0.7.0 - i.e. with an optional ID.

However, with a couple of @detro's points above, and also a  missing but crucial `else`, it works great:
```patch
diff --git a/schemamd/render.go b/schemamd/render.go
index c5e7db8..03af460 100644
--- a/schemamd/render.go
+++ b/schemamd/render.go
@@ -70,10 +70,6 @@ func writeAttribute(w io.Writer, path []string, att *tfjson.SchemaAttribute, gro
 		return nil, err
 	}
 
-	if name == "id" && att.Description == "" {
-		att.Description = "The ID of this resource."
-	}
-
 	if att.AttributeNestedType == nil {
 		err = WriteAttributeDescription(w, att, false)
 	} else {
@@ -225,14 +221,13 @@ nameLoop:
 			}
 		} else if childAtt, ok := block.Attributes[n]; ok {
 			for i, gf := range groupFilters {
-				if strings.ToLower(n) == "id" && childAtt.Description == "" {
-					if gf.topLevelTitle == "Read-Only" {
+				if strings.ToLower(n) == "id" {
+					if gf.topLevelTitle == "### Read-Only" {
 						childAtt.Description = "The ID of this resource."
 						groups[i] = append(groups[i], n)
 						continue nameLoop
 					}
-				}
-				if gf.filterAttribute(childAtt) {
+				} else if gf.filterAttribute(childAtt) {
 					groups[i] = append(groups[i], n)
 					continue nameLoop
 				}
```

Thanks again for this nmuesch. 🙂

OJFord commented on schemamd/render.go:
```diff
@@ -225,6 +225,13 @@ nameLoop:
+						groups[i] = append(groups[i], n)
+						continue nameLoop
+					}
+				}
 				if gf.filterAttribute(childAtt) {
```
```suggestion
				} else if gf.filterAttribute(childAtt) {
```

nmuesch commented:
Hey, apologies for my slow response. I did attempt to test things, but I've been a bit out of golang for a little while now and my test environment may have not been using the newly built version I was expecting. 

Anyway, thanks to both of you for continuing work on this and getting it past the finish line 🙇
